### ISSUE TITLE
Using --verifytypes on pandas GitHub repo

### ISSUE THREAD

**Issue by Dr-Irv** [microsoft/pyright#1383]
**Describe the bug**
To be able to work with an in-development version of pandas to improve the typing, I would need to be able to run `pyright` in my local development environment, which I set up using the instructions here: 
https://pandas.pydata.org/docs/development/contributing.html#creating-a-python-environment

Then if I have a directory `dev` with the `pandas` GitHub repo as a subfolder `pandas` and do `cd dev/pandas`, the folder `dev/pandas/pandas` has `__init__.py` and I can add a `py.typed` folder in there for testing.

But then I get this error when doing `pyright --verifytypes pandas`
```
  error: An internal error occurred while verifying types: "RangeError: Maximum call stack size exceeded
      at n._getLinesRecursive (C:\Anaconda3\node_modules\pyright\dist\pyright-internal.js:1:580656)
      at n._getLinesRecursive (C:\Anaconda3\node_modules\pyright\dist\pyright-internal.js:1:580666)
      at n._getLinesRecursive (C:\Anaconda3\node_modules\pyright\dist\pyright-internal.js:1:580666)
      at n.getString (C:\Anaconda3\node_modules\pyright\dist\pyright-internal.js:1:580141)
      at C:\Anaconda3\node_modules\pyright\dist\pyright-internal.js:1:165342
      at Map.forEach (<anonymous>)
      at y._verifySymbolsInSymbolTable (C:\Anaconda3\node_modules\pyright\dist\pyright-internal.js:1:164699)
      at y._verifyTypesForModule (C:\Anaconda3\node_modules\pyright\dist\pyright-internal.js:1:163706)
      at C:\Anaconda3\node_modules\pyright\dist\pyright-internal.js:1:161178
      at Array.forEach (<anonymous>)"
```

**To Reproduce**
This is on Windows.

Set up the pandas dev environment as described in the link above.
Then run `pyright --verifytypes pandas`

**Expected behavior**
No error

**VS Code extension or command-line**
Command line:  `pyright 1.1.102`

**Comment by erictraut** [comment]
I followed these instructions, and I'm not able to repro the problem you're seeing. Perhaps it's specific to Windows?

Have you tried deleting or renaming the "tests" directory so it is not included in the type verification?

**Comment by Dr-Irv** [comment]
> I followed these instructions, and I'm not able to repro the problem you're seeing. Perhaps it's specific to Windows?
> 

Could be.  But you're at Microsoft, so don't you use Windows?   :-)

> Have you tried deleting or renaming the "tests" directory so it is not included in the type verification?

Yes.

**Comment by erictraut** [comment]
We use many OSes at Microsoft :) I happen to do all of my development on Mac OS, but I have team members who use Windows and Linux as well.

Once you delete or rename the tests directory, does the internal error still occur?

**Comment by Dr-Irv** [comment]
> 
> Once you delete or rename the tests directory, does the internal error still occur?

Yes

**Comment by erictraut** [comment]
@Dr-Irv, just to confirm, this internal error occurs for you reliably when you follow the steps above _and_ you delete the "test" directory. Does it occur when you use the new feature that allows you to specify a submodule (e.g. "pyright --verifytypes pandas.core")? If not, can you try to narrow down the problem to a particular submodule?

I'm still not able to repro the problem on my machine, and without a repro, I don't think there's much I can do here.

**Comment by Dr-Irv** [comment]
@erictraut I definitely deleted the `tests` directory.

I can reproduce the error by typing:
```
pyright --verifytypes  pandas.core.reshape.concat
```
which yields:
```
Package name: "pandas.core.reshape.concat"
Package directory: "pandas"
Path of py.typed file: "pandas\py.typed"
  error: An internal error occurred while verifying types: "RangeError: Maximum call stack size exceeded
      at n._getLinesRecursive (C:\Anaconda3\node_modules\pyright\dist\pyright-internal.js:1:580265)
      at n.getString (C:\Anaconda3\node_modules\pyright\dist\pyright-internal.js:1:579750)
      at C:\Anaconda3\node_modules\pyright\dist\pyright-internal.js:1:165641
      at Map.forEach (<anonymous>)
      at y._verifySymbolsInSymbolTable (C:\Anaconda3\node_modules\pyright\dist\pyright-internal.js:1:164998)
      at y._verifyTypesForModule (C:\Anaconda3\node_modules\pyright\dist\pyright-internal.js:1:163972)
      at C:\Anaconda3\node_modules\pyright\dist\pyright-internal.js:1:161369
      at Array.forEach (<anonymous>)
      at y.verify (C:\Anaconda3\node_modules\pyright\dist\pyright-internal.js:1:161351)
      at C:\Anaconda3\node_modules\pyright\dist\pyright-internal.js:1:837626"

Public modules: 1
   pandas.core.reshape.concat (1 symbol)

Public symbols: 1
  Symbols with unknown type: 0
  Functions with missing docstring: 0
  Functions with missing default param: 0
  Classes with missing docstring: 0
Type completeness score: 100%

Completed in 2.158sec
```

**Comment by jakebailey** [comment]
Unrelated, but I'm a bit confused as to why the source maps aren't being loaded for the pyright CLI in the above. That trace is readable enough, but versions 1.1.94+ should all have loaded source maps...

**Comment by erictraut** [comment]
@Dr-Irv, yeah, there are a few mysteries here. Perhaps you're running an old version of node? If you type `node --version`, what does it say?

**Comment by jakebailey** [comment]
I'm working on testing this as closely as possible to the original report, will just take me a bit to get it set up. I think I just waited 10 minutes for conda to build the environment, only for it to fail right at the end due to my VS version being wrong...

**Comment by erictraut** [comment]
Yeah, it took me several tries to get it set up. I didn't use conda (I've had bad experiences with the conda installer messing up my environment in the past), so I used the venv approach instead. The instructions didn't specify which version of Python to use other than "it must be 3.6.1 or newer". I tried 3.6.7, 3.9, and 3.8. All of them failed to work because of incompatibilities with dependencies. I finally got 3.7 to work.

**Comment by jakebailey** [comment]
Crashed for me in exactly the same way:

```
(pandas-dev) E:\work\pandastesting\dev\pandas>python -c "import pandas; print(pandas.__version__)"
1.3.0.dev0+507.gb31c069de

(pandas-dev) E:\work\pandastesting\dev\pandas>pyright --version
pyright 1.1.103

(pandas-dev) E:\work\pandastesting\dev\pandas>pyright --verifytypes pandas.core.reshape.concat
Package name: "pandas.core.reshape.concat"
Package directory: "pandas"
Path of py.typed file: "pandas\py.typed"
  error: An internal error occurred while verifying types: "RangeError: Maximum call stack size exceeded
      at n._getLinesRecursive (C:\Users\jabaile\scoop\persist\nvm\nodejs\v12.19.0\node_modules\pyright\dist\pyright-internal.js:1:580265)
      at n.getString (C:\Users\jabaile\scoop\persist\nvm\nodejs\v12.19.0\node_modules\pyright\dist\pyright-internal.js:1:579750)
      at C:\Users\jabaile\scoop\persist\nvm\nodejs\v12.19.0\node_modules\pyright\dist\pyright-internal.js:1:165641
      at Map.forEach (<anonymous>)
      at y._verifySymbolsInSymbolTable (C:\Users\jabaile\scoop\persist\nvm\nodejs\v12.19.0\node_modules\pyright\dist\pyright-internal.js:1:164998)
      at y._verifyTypesForModule (C:\Users\jabaile\scoop\persist\nvm\nodejs\v12.19.0\node_modules\pyright\dist\pyright-internal.js:1:163972)
      at C:\Users\jabaile\scoop\persist\nvm\nodejs\v12.19.0\node_modules\pyright\dist\pyright-internal.js:1:161369
      at Array.forEach (<anonymous>)
      at y.verify (C:\Users\jabaile\scoop\persist\nvm\nodejs\v12.19.0\node_modules\pyright\dist\pyright-internal.js:1:161351)
      at C:\Users\jabaile\scoop\persist\nvm\nodejs\v12.19.0\node_modules\pyright\dist\pyright-internal.js:1:837626"

Public modules: 1
   pandas.core.reshape.concat (1 symbol)

Public symbols: 1
  Symbols with unknown type: 0
  Functions with missing docstring: 0
  Functions with missing default param: 0
  Classes with missing docstring: 0
Type completeness score: 100%

Completed in 2.541sec

(pandas-dev) E:\work\pandastesting\dev\pandas>
```

**Comment by jakebailey** [comment]
The source mapping thing is a red herring; even on my local build it doesn't appear for some reason. I'll have to look into that later.

**Comment by Dr-Irv** [comment]
> Crashed for me in exactly the same way:

Glad to see that you could finally reproduce it!!

**Comment by Dr-Irv** [comment]
Not sure if this will help, but if you edit `pandas/core/reshape/concat.py` and comment out the two overloads for `concat` at the top, then the error doesn't happen.

**Comment by Dr-Irv** [comment]
By the way, I did the `pyright --verifytypes pandas.core.reshape.concat` using a non-development version of pandas (i.e., clean environment, with pandas 1.2 installed, and creating a `py.typed` file), and got the same error.

**Comment by jakebailey** [comment]
I was able to reproduce this with plain pip install on Windows/Mac/Linux, so I don't think it's conda or Windows that's the problem here.

**Comment by jakebailey** [comment]
When debugging, I found that something (I think `pandas.core.reshape.concat.concat`) was so unknown that the diagnostic being produced to print out its "reason for being unknown" had over 300 thousand individual messages, and that the recursive guards weren't really kicking in because the tree was wide rather than deep along with how one of the guards was checked. Hopefully we can get Eric's machine to reproduce this, as I think he has a much better chance than me of figuring this one out.

I think this may just be exclusive to the type verification code, as the code can load otherwise; all of the messages are being constructed during verification after the type verifier has run. Maybe some recursion in that process that's being hit instead, but it fails at the end, not during construction of the large message.

**Comment by erictraut** [comment]
OK, I was able to repro the problem, and I understand what's happening. Strangely, it occurs only if you analyze certain deep submodules in isolation. If you analyze higher-level modules, it completes fine.

I've added a fix for the crash, although I'm going think more about better ways to handle the reporting of these issues.

**Comment by erictraut** [comment]
The crash is now addressed in Pyright 1.1.104, which I just published.

**Comment by Dr-Irv** [comment]
@erictraut thanks for this.  It now gets through the whole in-development pandas code, including the tests.
